module/vision/BallDetector/src/BallDetector.cpp

```diff
@@ -106,137 +108,167 @@ namespace module::vision {
 
-                balls->id        = horizon.id;
-                balls->timestamp = horizon.timestamp;
-                balls->Hcw       = horizon.Hcw;
+                balls->id        = horizon.id;         // camera id
+                balls->timestamp = horizon.timestamp;  // time when the image was taken
+                balls->Hcw       = horizon.Hcw;        // world to camera transform at the time the image was taken
+
+                // World to camera transform, to be used in for loop below
+                const Eigen::Affine3f Hcw(horizon.Hcw.cast<float>());
 
+                // CHECK EACH CLUSTER FOR VALID BALL
                 for (auto& cluster : clusters) {
                     Ball b;
+                    b.covariance = config.ball_angular_cov.asDiagonal();
 
-                    // Average the cluster to get the cones axis
-                    Eigen::Vector3f axis = Eigen::Vector3f::Zero();
+                    // FIND CENTRAL AXIS OF BALL
+                    // Add up all the unit vectors of each point (camera to point in world space) in the cluster to find
+                    // an average vector, which represents the central cone axis
+                    // uBCw: unit vector from camera to ball central axis in world space
+                    Eigen::Vector3f uBCw = Eigen::Vector3f::Zero();
                     for (const auto& idx : cluster) {
-                        axis += rays.col(idx);
+                        uBCw += uPCw.col(idx);
                     }
-                    axis /= cluster.size();
-                    axis.normalize();
+                    uBCw.normalize();  // get cone axis as a unit vector
 
-                    // Find the ray with the greatest distance from the axis
-                    // Should we use the average distance instead?
+                    // FIND UNIT RADIUS OF BALL
+                    // Find the ray (uPCw) with the greatest distance from the central axis (uBCw) to then determine the
+                    // largest radius possible from the edge points available. The radius is the upper bound on the ball
+                    // radius if it were a metre away. arccos(radius) is the angle between the central ball
+                    // axis (uBCw) and the edge of the ball. This helps to find the approximate distance to the ball
                     float radius = 1.0f;
                     for (const auto& idx : cluster) {
-                        const Eigen::Vector3f& ray(rays.col(idx));
-                        if (axis.dot(ray) < radius) {
-                            radius = axis.dot(ray);
-                        }
+                        // Unit vector from the camera to the ball edge, in world space
+                        const Eigen::Vector3f& uECw(uPCw.col(idx));
+                        // Choose this vector if it's further than previously seen vectors, to get the largest angle
+                        radius = uBCw.dot(uECw) < radius ? uBCw.dot(uECw) : radius;
```

Comment:
You might want to say that we are checking `<` here because `radius` is actually `cos(theta)` (the `cos` of the angle, `theta`, between the axis and the ball edge) and larger `theta` results in smaller `cos(theta)`. This logic can become non-intuitive when you are talking about finding the "largest" angle